### PR TITLE
When pleg channel is full, discard events and record its count

### LINKED ISSUE #72482: Deadlock in PLEG relist() for health check and Kubelet syncLoop()
**What happened**:

In our stress testing, we found the deadlock issue in PLEG relist() for healthy checking and Kubelet syncLoop()

The goroutine stack trace is as following

```
goroutine 386 [chan send, 1140 minutes]:
k8s.io/kubernetes/pkg/kubelet/pleg.(*GenericPLEG).relist(0xc42069ea20)
	/workspace/anago-v1.11.5-beta.0.24+753b2dbc622f5c/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pleg/generic.go:261 +0x74e
k8s.io/kubernetes/pkg/kubelet/pleg.(*GenericPLEG).(k8s.io/kubernetes/pkg/kubelet/pleg.relist)-fm()
	/workspace/anago-v1.11.5-beta.0.24+753b2dbc622f5c/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pleg/generic.go:130 +0x2a
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc4212ee520)
	/workspace/anago-v1.11.5-beta.0.24+753b2dbc622f5c/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc4212ee520, 0x3b9aca00, 0x0, 0x1, 0xc420056540)
	/workspace/anago-v1.11.5-beta.0.24+753b2dbc622f5c/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc4212ee520, 0x3b9aca00, 0xc420056540)
	/workspace/anago-v1.11.5-beta.0.24+753b2dbc622f5c/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by k8s.io/kubernetes/pkg/kubelet/pleg.(*GenericPLEG).Start
	/workspace/anago-v1.11.5-beta.0.24+753b2dbc622f5c/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pleg/generic.go:130 +0x88
...
goroutine 309 [sleep]:
time.Sleep(0x12a05f200)
	/usr/local/go/src/runtime/time.go:102 +0x166
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncLoop(0xc4205e3b00, 0xc420ff2780, 0x3e56a60, 0xc4205e3b00)
	/workspace/anago-v1.11.5-beta.0.24+753b2dbc622f5c/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1777 +0x1e7
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).Run(0xc4205e3b00, 0xc420ff2780)
	/workspace/anago-v1.11.5-beta.0.24+753b2dbc622f5c/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1396 +0x27f
k8s.io/kubernetes/cmd/kubelet/app.startKubelet.func1()
	/workspace/anago-v1.11.5-beta.0.24+753b2dbc622f5c/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:998 +0x67
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc42105dfb0)
	/workspace/anago-v1.11.5-beta.0.24+753b2dbc622f5c/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc42105dfb0, 0x0, 0x0, 0x1, 0xc420056540)
	/workspace/anago-v1.11.5-beta.0.24+753b2dbc622f5c/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc42105dfb0, 0x0, 0xc420056540)
	/workspace/anago-v1.11.5-beta.0.24+753b2dbc622f5c/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by k8s.io/kubernetes/cmd/kubelet/app.startKubelet
	/workspace/anago-v1.11.5-beta.0.24+753b2dbc622f5c/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/kubelet/app/server.go:996 +0xea
...
```

We did some investigation, and found the following issue.

```
...
func (kl *Kubelet) syncLoop(updates <-chan kubetypes.PodUpdate, handler SyncHandler) {
...
	for {
		if rs := kl.runtimeState.runtimeErrors(); len(rs) != 0 {
			glog.Infof("skipping pod synchronization - %v", rs)
			// exponential backoff
			time.Sleep(duration)
			duration = time.Duration(math.Min(float64(max), factor*float64(duration)))
			continue
		}
		// reset backoff if we have a success
		duration = base

		kl.syncLoopMonitor.Store(kl.clock.Now())
		if !kl.syncLoopIteration(updates, handler, syncTicker.C, housekeepingTicker.C, plegCh) {
			break
		}
		kl.syncLoopMonitor.Store(kl.clock.Now())
	}
}

...
func (kl *Kubelet) syncLoopIteration(configCh <-chan kubetypes.PodUpdate, handler SyncHandler,
	syncCh <-chan time.Time, housekeepingCh <-chan time.Time, plegCh <-chan *pleg.PodLifecycleEvent) bool {
	select {
         ...
	case e := <-plegCh:
		if isSyncPodWorthy(e) {
			// PLEG event for a pod; sync it.
			if pod, ok := kl.podManager.GetPodByUID(e.ID); ok {
				glog.V(2).Infof("SyncLoop (PLEG): %q, event: %#v", format.Pod(pod), e)
				handler.HandlePodSyncs([]*v1.Pod{pod})
			} else {
				// If the pod no longer exists, ignore the event.
				glog.V(4).Infof("SyncLoop (PLEG): ignore irrelevant event: %#v", e)
			}
		}

		if e.Type == pleg.ContainerDied {
			if containerID, ok := e.Data.(string); ok {
				kl.cleanUpContainersInPod(e.ID, containerID)
			}
		}
         ...
	}
	return true
}
```

The main loop for event processing in Kubelet will check if there are some runtimeErrors: if yes, wait until runtime back to normal; if no, it will call syncLoopIteration to read and process events. 

And the runtime healthChecks depends on the GenericPLEG.relist. But in the relist() impl, it will send the pod lifecycle event to eventChannel which may introduce deadlock when eventChannel is full. 

```
func (g *GenericPLEG) relist() {
        ...
	// If there are events associated with a pod, we should update the
	// podCache.
	for pid, events := range eventsByPodID {
                 ... 
		for i := range events {
			// Filter out events that are not reliable and no other components use yet.
			if events[i].Type == ContainerChanged {
				continue
			}
			g.eventChannel <- events[i]
		}
	}

	... 
}

```

And in the above code, the eventChannel will not be consumed if the runtime is not healthy. In some situation, the eventChannel may full, and the relist() will hang in sending new pod lifecycle event. And the runtime status will not be able to recovered. 

It is related to issues #45419

**What you expected to happen**:  

No such issue

**How to reproduce it (as minimally and precisely as possible)**:

It is not easy to reproduce, it happens after few days stress testing (scaling the deployment repeatedly with high load). 

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.11.5
- Cloud provider or hardware configuration: Bare metal
- OS (e.g. from /etc/os-release):  CentOS 7.4
- Kernel (e.g. `uname -a`):
Linux bjc-qtt-k8s-node-010 3.10.0-693.2.2.el7.x86_64 #1 SMP Tue Sep 12 22:26:13 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: kubeadm
- Others:

/kind bug

### PR DESCRIPTION
**Pull request by changyaowei**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind bug
> /kind design

**What this PR does / why we need it**:
When event channel is full, discard the event, if continue put event in channel and block a `relistThreshold` time,  `pleg Healthy()` will return error, if like so, `kubelet syncloop()` get a runtime error and not to consume event, it will lead to a deadlock

image

In `runtimeErrors func` invoke `pleg health func`
image

image

**Which issue(s) this PR fixes**:

Fixes #72482 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
when pleg channel is full, discard events and record its count
```

### CODE REVIEW

File: pkg/kubelet/pleg/generic.go [hunk 1]
```diff
@@ -192,6 +192,12 @@ func (g *GenericPLEG) relist() {
 		metrics.PLEGRelistInterval.Observe(metrics.SinceInMicroseconds(lastRelistTime))
 	}
 
+	_, err := g.runtime.Status()
+	if err != nil {
+		klog.Errorf("Container runtime sanity check failed: %v", err)
```
Comment (resouer) [inline pkg/kubelet/pleg/generic.go]: Let's say the container runtime is dead, and PLEG returned here with logged this line of error. Then what will happen next? 

What will be the status of Pods? And how user knows the container runtime is dead? (user will never check log unless sth really bad happens)

Comment (changyaowei) [reply pkg/kubelet/pleg/generic.go]: Container runtime dead not represent container dead, container will not be impacted. We return here, 3 minutes later, the node status will be not ready which we get from apiserver.

File: pkg/kubelet/pleg/generic.go [hunk 2]
```diff
@@ -224,6 +224,15 @@ func (g *GenericPLEG) relist() {
 		}
 	}
 
+	newEventCount := 0
+	for _, value := range eventsByPodID {
+		newEventCount = newEventCount + len(value)
+	}
+	if (newEventCount + len(g.eventChannel)) > cap(g.eventChannel) {
+		klog.Errorf("new event count %d, now event channel is %d, cap is %s, channel will full, so discard",newEventCount,len(g.eventChannel),cap(g.eventChannel))
```
Comment (resouer) [inline pkg/kubelet/pleg/generic.go]: We don't need to reveal the accurate number of events which will is not quite helpful to debug, so simply:

```suggestion
		klog.Errorf("event channel will be full after this relist cycle, discard",newEventCount,len(g.eventChannel),cap(g.eventChannel))
```

File: pkg/kubelet/pleg/generic.go [hunk 3]
```diff
@@ -224,6 +224,15 @@ func (g *GenericPLEG) relist() {
 		}
 	}
 
+	newEventCount := 0
```
Comment (resouer) [inline pkg/kubelet/pleg/generic.go]: I will prefer: `upcomingEventCount`

**Comment by resouer** [comment]
Please update the PR title, and mark it as WIP until proper tests are added.

**Comment by resouer** [comment]
ref #72482

File: pkg/kubelet/pleg/generic.go (hunk 2 above)
Comment (changyaowei) [reply pkg/kubelet/pleg/generic.go]: OK

File: pkg/kubelet/pleg/generic.go (hunk 3 above)
Comment (changyaowei) [reply pkg/kubelet/pleg/generic.go]: ok

File: pkg/kubelet/pleg/generic_test.go [hunk 4]
```diff
@@ -42,15 +42,21 @@ type TestGenericPLEG struct {
 	clock   *clock.FakeClock
 }
 
-func newTestGenericPLEG() *TestGenericPLEG {
+func newTestGenericPLEGWithLargeCap() *TestGenericPLEG {
```
Comment (resouer) [inline pkg/kubelet/pleg/generic_test.go]: nits: `newTestGenericPLEGWithLargeChannel`

File: pkg/kubelet/pleg/generic_test.go [hunk 5]
```diff
@@ -42,15 +42,21 @@ type TestGenericPLEG struct {
...
+	// The channel capacity should be large enough to hold all events in a
+	// single test.
+	return newTestGenericPLEG(100)
```
Comment (resouer) [inline pkg/kubelet/pleg/generic_test.go]: Make it a constant

File: pkg/kubelet/pleg/generic_test.go [hunk 6]
```diff
@@ -157,6 +163,85 @@ func TestRelisting(t *testing.T) {
 	verifyEvents(t, expected, actual)
 }
 
+func TestEventChannelFull(t *testing.T) {
```
Comment (resouer) [inline pkg/kubelet/pleg/generic_test.go]: We need to add a comment here to explain why this test is needed :-)

**Comment by yujuhong** [comment]
This PR deserves a release note, so please add one :-)

The current fix does not update PLEG's internal cache, so the next relist cycle, PLEG will still try to send all the events.
The downside is that, if kubelet's overwhelmed for a long duration, the events may accumulate to a point that they can never be sent out all together (#events > channel capacity).

Another approach is to update the internal cache AND drop the events, then relying on the kubelet's fallback re-sync (default: 1 min) to catch up. This may mean longer delay for a pod to sync at times, but it will not suffer the same problem as the current fix.

Also, I don't think kubelet should be operating in this condition perpetually. I'd suggest adding a metrics to record how many events have been dropped. There are already some example of collecting metrics in PLEG, e.g., see `PLEGRelistLatency`.

**Comment by resouer** [comment]
Thanks for quick review!

> Another approach is to update the internal cache AND drop the events, then relying on the kubelet's fallback re-sync (default: 1 min) to catch up

I believe 1 min gap is fine in this case.

File: pkg/kubelet/pleg/generic.go [hunk 7]
```diff
@@ -224,6 +224,15 @@ func (g *GenericPLEG) relist() {
...
 	}
 
+	upcomingEventCount := 0
+	for _, value := range eventsByPodID {
+		upcomingEventCount = upcomingEventCount + len(value)
+	}
+	if (upcomingEventCount + len(g.eventChannel)) > cap(g.eventChannel) {
+		klog.Error("event channel is about to be full, discard this relist() cycle")
+		return
+	}
```
Comment (Random-Liu) [inline pkg/kubelet/pleg/generic.go]: Don't do this. Just:
```
select {
case g.eventChannel <- events[i]:
default:
  // log skipping event
}
```

Comment (changyaowei) [reply pkg/kubelet/pleg/generic.go]: If we do this , we should update `g.podRecords`, and podRecord.old  will update to current pod. In next relist(), the events which get from compare `g.podRecords.current` and `g.podRecords.old` may be lost.

**Comment by changyaowei** [comment]
> relying on the kubelet's fallback re-sync (default: 1 min) to catch up.

@yujuhong Sorry  I can’t not understand which re-sync do you mean, can you explain it ? or screenshot the code, thanks

**Comment by resouer** [comment]
> If we do this , we should update g.podRecords, and podRecord.old will update to current pod. In next relist(), the events which get from compare g.podRecords.current and g.podRecords.old may be lost.

> relying on the kubelet's fallback re-sync (default: 1 min) to catch up.

@changyaowei We are referring to relying on kubelet's main sync loop to recover data in next cycle, i.e. 1 min later.

So you can just skip the event (no need to return) and let the PLEG cache been updated as is.

File: pkg/kubelet/pleg/generic_test.go [hunk 8]
```diff
@@ -157,6 +167,89 @@ func TestRelisting(t *testing.T) {
 	verifyEvents(t, expected, actual)
 }
 
+// TestEventChannelFull test when channel is full, the events will be discard.
+func TestEventChannelFull(t *testing.T) {
+	prometheus.MustRegister(metrics.PLEGDiscardEvents)
```
Comment (resouer) [inline pkg/kubelet/pleg/generic_test.go]: Do not rely on prometheus to do unit testing.

**Review by Random-Liu — COMMENTED** [review]
LGTM overall with one comment

**Comment by redbaron** [comment]
Does it qualify for cherry pick into 1.11 and 1.12?

**Comment by changyaowei** [comment]
@Random-Liu Does it need to cherry pick into 1.11, 1.12 and 1.13?

**Comment by apelisse** [comment]
Either the feature doesn't work or the test is broken, but `TestEventChannelFull` is extremely flaky.

**Comment by apelisse** [comment]
```
//pkg/kubelet/pleg:go_default_test                                       FAILED in 230 out of 1000 in 0.2s
```

**Comment by changyaowei** [comment]
@apelisse By test , i found that the  line of ` klog.Error` take a long time.

**Comment by apelisse** [comment]
Are you saying that this is a timing issue? In general, time shouldn't have an impact in unit-tests. I don't have any problem with the time, my main concern is the flakiness.

**Comment by changyaowei** [comment]
@apelisse I run the test in my mac and everything is ok, and can you give your fail output
image

**Comment by apelisse** [comment]
Yeah, you ran it twice, I ran it 1000 times :-) Also note that go caches the output after the first success, so you can't simple run it multiple times naively.

It failed 230 times out of the 1000 times I ran it.

**Comment by resouer** [comment]
@apelisse We are trying to reproduce and dig into the change now.

**Comment by apelisse** [comment]
You can reproduce by running:
```
bazel test --runs_per_test=1000 //pkg/kubelet/pleg:go_default_test
```

**Comment by ping035627** [comment]
@apelisse, @changyaowei
How has this problem progressed? TestEventChannelFull is also often fails in my ci environment.

**Comment by gsjeon** [comment]
@changyaowei Does it qualify for cherry pick into 1.12 ?

**Pull request merged**